File: backend/src/script/lib/agn3/fsdb.py
####################################################################################################################################################################################################################################################################
#                                                                                                                                                                                                                                                                  #
#                                                                                                                                                                                                                                                                  #
#        Copyright (C) 2022 AGNITAS AG (https://www.agnitas.org)                                                                                                                                                                                                   #
#                                                                                                                                                                                                                                                                  #
#        This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.    #
#        This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.           #
#        You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.                                                                                                            #
#                                                                                                                                                                                                                                                                  #
####################################################################################################################################################################################################################################################################
#
from	__future__ import annotations
import	os, stat
from	datetime import datetime
from	types import TracebackType
from	typing import Optional
from	typing import Dict, Generator, NamedTuple, Type
from	.definitions import base
from	.ignore import Ignore
from	.io import create_path
from	.tools import calc_hash
#
__all__ = ['FSDB']
#
class FSDB:
	__slots__ = ['cache']
	class Value (NamedTuple):
		content: str
		timestamp: datetime

	path = os.path.join (base, 'var', 'fsdb')
	def __init__ (self) -> None:
		self.cache: Dict[str, Optional[FSDB.Value]] = {}

	def clear (self) -> None:
		self.cache.clear ()
	
	def get (self, key: str, default: Optional[str] = None) -> Optional[str]:
		try:
			return self[key]
		except KeyError:
			return default

	def __enter__ (self) -> FSDB:
		return self

	def __exit__ (self, exc_type: Optional[Type[BaseException]], exc_value: Optional[BaseException], traceback: Optional[TracebackType]) -> Optional[bool]:
		return None

	def __iter__ (self) -> Generator[str, None, None]:
		depth = [FSDB.path]
		while depth:
			current = depth.pop (0)
			with Ignore (OSError):
				for filename in sorted (os.listdir (current)):
					path = os.path.join (current, filename)
					st = os.stat (path)
					if stat.S_ISDIR (st.st_mode):
						depth.append (path)
					elif stat.S_ISREG (st.st_mode):
						key = self.__convert_path_to_key (path)
						if key is not None:
							yield key

	def __getitem__ (self, key: str) -> str:
		return self.retrieve (key).content
	
	def __setitem__ (self, key: str, value: str) -> None:
		self.store (key, value)
	
	def __delitem__ (self, key: str) -> None:
		self.remove (key)

	def retrieve (self, key: str) -> FSDB.Value:
		with Ignore (KeyError):
			value = self.cache[key]
			if value is None:
				raise KeyError (key)
			return value
		path = self.__convert_key_to_path (key)
		if path is not None:
			with Ignore (IOError, UnicodeDecodeError), open (path) as fd:
				content = fd.read ()
				try:
					timestamp = datetime.fromtimestamp (os.stat (fd.fileno ()).st_mtime)
				except OSError:
					timestamp = datetime.now ()
				value = self.cache[key] = FSDB.Value (content, timestamp)
				return value
		self.cache[key] = None
		raise KeyError (key)
	
	def store (self, key: str, content: str) -> None:
		path = self.__convert_key_to_path (key)
		if path is not None:
			with Ignore (KeyError):
				del self.cache[key]
			try:
				with Ignore (IOError, UnicodeDecodeError), open (path) as fd:
					if fd.read () == content:
						with Ignore (OSError):
							os.utime (fd.fileno ())
						return
				try:
					with open (path, 'w') as fd:
						fd.write (content)
				except FileNotFoundError:
					create_path (os.path.dirname (path))
					with open (path, 'w') as fd:
						fd.write (content)
			finally:
				try:
					timestamp = datetime.fromtimestamp (os.stat (path).st_mtime)
				except OSError:
					timestamp = datetime.now ()
				self.cache[key] = FSDB.Value (content, timestamp)
		else:
			raise ValueError (f'{key}: invalid key')
	
	def remove (self, key: str) -> None:
		path = self.__convert_key_to_path (key)
		if path is not None:
			with Ignore (KeyError):
				del self.cache[key]
			with Ignore (OSError):
				os.unlink (path)
		else:
			raise ValueError (f'{key}: invalid key')

	hash_chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_-'
	def __convert_key_to_path (self, key: str) -> Optional[str]:
		if os.path.sep in key:
			return None
		elements = key.split (':')
		elements.insert (-1, self.__hash_representation (elements[-1]))
		return os.path.join (FSDB.path, os.path.join (*elements))

	def __convert_path_to_key (self, path: str) -> Optional[str]:
		if path.startswith (FSDB.path):
			elements = path[-(len (FSDB.path) - 2):].split ('/')
			with Ignore (UnicodeEncodeError):
				if len (elements) > 1 and elements.pop (-2) == self.__hash_representation (elements[-1]):
					return ':'.join (elements)
		return None

	def __hash_representation (self, name: str) -> str:
		hash_value = calc_hash (name)
		return FSDB.hash_chars[(hash_value >> 6) % len (FSDB.hash_chars)] + FSDB.hash_chars[hash_value % len (FSDB.hash_chars)]

File: backend/src/c/lib/fsdb.c
/********************************************************************************************************************************************************************************************************************************************************************
 *                                                                                                                                                                                                                                                                  *
 *                                                                                                                                                                                                                                                                  *
 *        Copyright (C) 2019 AGNITAS AG (https://www.agnitas.org)                                                                                                                                                                                                   *
 *                                                                                                                                                                                                                                                                  *
 *        This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.    *
 *        This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.           *
 *        You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.                                                                                                            *
 *                                                                                                                                                                                                                                                                  *
 ********************************************************************************************************************************************************************************************************************************************************************/
# include	<stdlib.h>
# include	<fcntl.h>
# include	<errno.h>
# include	<sys/types.h>
# include	<sys/stat.h>
# include	<utime.h>
# include	"agn.h"

# define	HASH_CHARS		"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_-"

struct fsdb { /*{{{*/
	char	path[PATH_MAX + 1];	/* store the base path and is used to build final path	*/
	char	*fptr;			/* pointer to the variable part of path			*/
	int	base_path_len;		/* length of base path with trailing /			*/
	int	fsize;			/* free size in path					*/
	char	scratch[PATH_MAX + 1];	/* for creating temp. files				*/
	char	*sptr;			/* pointer to the variable part of scratch		*/
	/*}}}*/
};

static bool_t
convert_key (fsdb_t *f, const char *key) /*{{{*/
{
	int		keylen = strlen (key);
	const char	*src, *current;
	char		*dst;
	hash_t		hash;
	
	if ((keylen + 5 > f -> fsize) || strchr (key, '/'))
		return false;
	for (src = key, dst = f -> fptr; src; ) {
		current = src;
		src = strchr (src, ':');
		if (src) {
			if (src - current) {
				memcpy (dst, current, src - current);
				dst += src - current;
			}
			*dst++ = '/';
			++src;
		} else {
			hash = hash_svalue (current, strlen (current), false);
			*dst++ = HASH_CHARS[(hash >> 6) % (sizeof (HASH_CHARS) - 1)];
			*dst++ = HASH_CHARS[hash % (sizeof (HASH_CHARS) - 1)];
			*dst++ = '/';
			strcpy (dst, current);
		}
	}
	return true;
}/*}}}*/
static bool_t
create_path (fsdb_t *f) /*{{{*/
{
	char		temp[PATH_MAX + 1];
	char		*ptr;
	struct stat	st;
	
	strncpy (temp, f -> path, PATH_MAX);
	temp[PATH_MAX] = '\0';
	if (ptr = strrchr (temp, '/'))
		*ptr = '\0';
	ptr = temp + f -> base_path_len;
	while (ptr) {
		if (ptr = strchr (ptr, '/'))
			*ptr = '\0';
		if (stat (temp, & st) != -1) {
			if (! S_ISDIR (st.st_mode))
				break;
		} else {
			if ((mkdir (temp, 0700) == -1) && ((stat (temp, & st) == -1) || (! S_ISDIR (st.st_mode))))
				break;
		}
		if (ptr)
			*ptr++ = '/';
		else
			return true;
	}
	return false;
}/*}}}*/
static bool_t
do_read (int fd, void *buffer, int size) /*{{{*/
{
	int	n;
	
	while (size > 0) {
		if ((n = read (fd, buffer, size)) > 0) {
			buffer = (char *) buffer + size;
			size -= n;
		} else
			break;
	}
	return size == 0 ? true : false;
}/*}}}*/
static bool_t
do_write (int fd, const void *buffer, int size) /*{{{*/
{
	int	n;
	
	while (size > 0) {
		if ((n = write (fd, buffer, size)) > 0) {
			buffer = (char *) buffer + size;
			size -= n;
		} else
			break;
	}
	return size == 0 ? true : false;
}/*}}}*/
fsdb_result_t *
fsdb_result_alloc (int size) /*{{{*/
{
	fsdb_result_t	*r;
	
	if (r = (fsdb_result_t *) malloc (sizeof (fsdb_result_t)))
		if (r -> value = malloc (size + 1))
			r -> vlen = size;
		else {
			free (r);
			r = NULL;
		}
	return r;
}/*}}}*/
fsdb_result_t *
fsdb_result_free (fsdb_result_t *r) /*{{{*/
{
	if (r) {
		if (r -> value)
			free (r -> value);
		free (r);
	}
	return NULL;
}/*}}}*/
fsdb_t *
fsdb_alloc (const char *base_path) /*{{{*/
{
	char	temp[PATH_MAX + 1];
	int	bp_len;
	fsdb_t	*f;
	
	if (! base_path) {
		if (snprintf (temp, sizeof (temp) - 1, "%s/var/fsdb/", path_home ()) == -1)
			return NULL;
		base_path = temp;
	}
	bp_len = strlen (base_path);
	f = NULL;
	if ((bp_len + 256 < PATH_MAX) && (f = (fsdb_t *) malloc (sizeof (fsdb_t)))) {
		strncpy (f -> path, base_path, bp_len);
		f -> fptr = f -> path + bp_len;
		if (*(f -> fptr) != '/')
			*(f -> fptr++) = '/';
		f -> base_path_len = f -> fptr - f -> path;
		f -> fsize = PATH_MAX - f -> base_path_len;
		strncpy (f -> scratch, f -> path, f -> base_path_len);
		f -> sptr = f -> scratch + f -> base_path_len;
	}
	return f;
}/*}}}*/
fsdb_t *
fsdb_free (fsdb_t *f) /*{{{*/
{
	if (f) {
		free (f);
	}
	return NULL;
}/*}}}*/
bool_t
fsdb_exists (fsdb_t *f, const char *key) /*{{{*/
{
	struct stat	st;
	
	if (convert_key (f, key) && (stat (f -> path, & st) != -1) && S_ISREG (st.st_mode))
		return true;
	return false;
}/*}}}*/
fsdb_result_t *
fsdb_get (fsdb_t *f, const char *key) /*{{{*/
{
	fsdb_result_t	*r = NULL;
	int		fd;
	
	if (convert_key (f, key) && ((fd = open (f -> path, O_RDONLY)) != -1)) {
		struct stat	st;
		
		if ((fstat (fd, & st) != -1) && (r = fsdb_result_alloc (st.st_size))) {
			if (do_read (fd, r -> value, r -> vlen))
				((char *) r -> value)[r -> vlen] = '\0';
			else
				r = fsdb_result_free (r);
		}
		close (fd);
	}
	return r;
}/*}}}*/
bool_t
fsdb_put (fsdb_t *f, const char *key, const void *value, int vlen) /*{{{*/
{
	bool_t	rc = false;
	
	if (convert_key (f, key)) {
		int	fd;
		
		if ((fd = open (f -> path, O_RDONLY)) != -1) {
			struct stat	st;
			void		*temp;
			
			if ((fstat (fd, & st) != -1) && (st.st_size == vlen) && (temp = malloc (st.st_size))) {
				if (do_read (fd, temp, vlen))
					if ((! vlen) || (! memcmp (value, temp, vlen))) {
						struct utimbuf	utbuf;
						
						utbuf.actime = utbuf.modtime = time (NULL);
						if (utime (f -> path, & utbuf) != -1)
							rc = true;
					}
				free (temp);
			}
			close (fd);
		}
		if (! rc) {
			snprintf (f -> sptr, f -> base_path_len, ".%06d.%10ld", getpid (), time (NULL));
			if ((fd = open (f -> scratch, O_WRONLY | O_CREAT | O_TRUNC, 0600)) != -1) {
				if (do_write (fd, value, vlen)) {
					if ((rename (f -> scratch, f -> path) != -1) ||
					    (create_path (f) && (rename (f -> scratch, f -> path) != -1)))
						rc = true;
				}
				close (fd);
				unlink (f -> scratch);
			}
		}
	}
	return rc;
}/*}}}*/
bool_t
fsdb_remove (fsdb_t *f, const char *key) /*{{{*/
{
	if (convert_key (f, key)) {
		struct stat	st;
		
		if (stat (f -> path, & st) == -1) {
			if (errno == ENOENT)
				return true;
		} else if (S_ISREG (st.st_mode) && (unlink (f -> path) != -1))
				return true;
	}
	return false;
}/*}}}*/
